### DISCUSSION TITLE
onKeyboardNavigate for useListNavigation?

### DISCUSSION THREAD

**Discussion by what-is-a-crow** [floating-ui/floating-ui#1981] [React]
Our select component is powered by `useListNavigation` and we are using `onNavigate` to set an `activeIndex` that is checked against an option's index to change its visual state (in our case, our global keyboard-focus style). This works but means that clicking on an option (versus selecting with the keyboard) applies this keyboard-focus style, which is undesirable.

Is there a way to know whether or not a selection action was triggered by keyboard versus a mouse click? If not, would adding an `onKeyboardNavigate` property be feasible?

**Comment by atomiks** [comment]
How does it receive the global keyboard focus style upon click? The `focusItemOnHover` option (happens before click) or the browser focusing the list item element natively, triggering `:focus` styling? I guess you can/should use `:focus-visible`?

**Comment by what-is-a-crow** [comment]
Clicking an option calls `onNavigate`, which then sets `activeIndex`, which we're using to show which option is currently focused.

We can't use `:focus-visible` because the trigger element is often the one with the _actual_ focus. (Worth calling out that we are aware of the combobox API, but that didn't end up being sufficient for our use case- it's been long enough now that I don't recall what the problems were.)

**Reply by atomiks** [reply]
> Clicking an option calls onNavigate, which then sets activeIndex

Why does this happen? I am assuming you're using `virtual: true` since the trigger element has real DOM focus? Clicking shouldn't have an effect on the active item id

I'm guessing you need to make the virtual items non-focusable elements, like a regular `<div>` to prevent focus styling from being triggered on them

**Reply by what-is-a-crow** [reply]
Aha- I just assumed that was expected behavior (clicking triggering `onNavigate`), but it looks like it's actually `onFocus`, which is being triggered on click. If I short circuit that behavior (`e.stopPropagation()` for `onFocus` on my option node), it works- keyboard nav shows properly and clicking never triggers the active index. Is this a hack?
